borsh/src/de/mod.rs

```diff
@@ -294,6 +294,57 @@ impl BorshDeserialize for String {
     }
 }
 
+#[cfg(feature = "bigdecimal")]
+impl BorshDeserialize for bigdecimal::BigDecimal {
+    #[inline]
+    fn deserialize(buf: &mut &[u8]) -> Result<Self> {
+        let digits = bigdecimal::num_bigint::BigInt::deserialize(buf)?;
+        let scale = i64::deserialize(buf)?;
+        Ok(bigdecimal::BigDecimal::new(digits, scale))
+    }
+}
+
+#[cfg(feature = "bigdecimal")]
+impl BorshDeserialize for bigdecimal::num_bigint::BigInt {
+    #[inline]
+    fn deserialize(buf: &mut &[u8]) -> Result<Self> {
+        let sign = bigdecimal::num_bigint::Sign::deserialize(buf)?;
+        let digits = <Vec<u8>>::deserialize(buf)?;
+        assert!(
+            sign == bigdecimal::num_bigint::Sign::NoSign
+                || (!digits.is_empty() && digits.last().unwrap() > &0)
```

Comment:
This check still allows for non-canonical deserialization if the sign is `NoSign` since in this case, it doesn't check for the padded bytes

Test case I used to verify:
```
	assert_eq!(BigInt::try_from_slice(&[1,1,0,0,0,0]).unwrap(), BigInt::default());
    assert_eq!(BigInt::try_from_slice(&[1,2,0,0,0,0,0]).unwrap(), BigInt::default());
```

Also, side note, wondering why unsigned varints aren't used for the byle length? Seems weird we are just encoding a u32 (4 bytes) for the length even when it's small? Given most size indicators are small and at max it would be 5 bytes, seems like it would be more compact in most cases? Was this done to minimize complexity and code size rather than serialization size? cc @matklad @frol

Reply:
Only the last one failed; added this test.

Reply:
but only one should pass, though. We need a canonical representation of 0. I think it should be `[]` and not `[0]` for compactness

Reply:
Missed the Q about varints, sorry. I don't know the motivation for using u32 rather than varints for lengths. Abstractly, I think I agree that varint would've be a better choice, as that would be more compact, and also would get rid of an error condition when serializing vectors and such. 


However, we already have a precedent that we use `u32` to encode container lengths, and I think it's important that we use the same repr for bigints. 

That is, using varints *is* more efficient for bigints, but this isn't an important optimiation, because there are far more array & string lengths than bigint lengths. 

OTOH, doing something *different* for bigints adds complexity cost